File: internal/sem/testdata/fixtures/php-oo/Shapes.php
<?php

namespace Shapes;

interface Shape
{
    public function area(): float;
}

abstract class Base
{
    abstract public function describe(): string;
}

class Circle extends Base implements Shape
{
    public function area(): float
    {
        return 3.14;
    }

    public function describe(): string
    {
        return "circle";
    }

    // PHP has no nested function scope: running scaled() declares `area` in the
    // CURRENT NAMESPACE, after which it is the plain function `Shapes\area()`.
    // `Circle::area` is the real member above and calling it as a method is a
    // fatal error, so this must emit as the function `Circle.scaled.area` —
    // emitting the method `Circle.area` invents a second member and pushes the
    // real one onto a `#sig:`-disambiguated ID (issue #259).
    public function scaled(float $factor): float
    {
        function area(float $radius): float
        {
            return 3.14 * $radius;
        }

        return area($factor);
    }
}

File: internal/sem/testdata/fixtures/rust-oo/shapes.rs
pub trait Shape {
    fn area(&self) -> f64;
}

pub trait Drawable: Shape {
    fn draw(&self);
}

pub struct Circle {
    pub radius: f64,
}

impl Shape for Circle {
    fn area(&self) -> f64 {
        3.14 * self.radius * self.radius
    }
}

impl Drawable for Circle {
    fn draw(&self) {}
}

impl Circle {
    // A `fn` declared inside a method body is an ITEM scoped to that BLOCK, not
    // an associated function of Circle (issue #259). It takes no `self` and
    // `Circle::area` resolves only to the trait impl above, so this must emit
    // as the function `Circle.scaled.area` — emitting the method `Circle.area`
    // would name a second, non-existent associated function and push the real
    // one onto a `#sig:`-disambiguated ID.
    pub fn scaled(&self, factor: f64) -> f64 {
        fn area(radius: f64) -> f64 {
            3.14 * radius * radius
        }

        area(self.radius) * factor
    }

    // A same-named nested item in a DIFFERENT method must stay distinct rather
    // than collapsing onto one ID.
    pub fn halved(&self) -> f64 {
        fn area(radius: f64) -> f64 {
            1.57 * radius * radius
        }

        area(self.radius)
    }
}
